hack/lib/log.sh

```diff
@@ -0,0 +1,216 @@
+# This function needs the non-data columns of the file to be prefixed with comments. 
+#
+# Globals:
+#  - LOG_DIR
+# Arguments:
+#  - 1: data file
+# Returns:
+#  None
+function os::log::internal::plot() {
+    local datafile=$1
+    local plotname
+    plotname="$(basename "${datafile}" .txt)"
+
+    # we are expecting the output of a 'sadf -d' read, so the headers will be on the first line of the file
+    local headers
+    headers=( $( head -n 1 "${datafile}" | sed 's/^# //g' | tr ';' ' ' ) )
+
+    local records
+    local width
+    records="$(( $( cat "${datafile}" | wc -l ) - 1 ))" # one of these lines will be the header comment
+    if [[ "${records}" -gt 90 ]]; then
+        width="$(echo "8.5 + ${records}*0.025" | bc )"
+    else 
+        width="8.5"
+    fi
+
+    local gnuplot_directive=( "set terminal pdf size ${width}in,$(( 2 * (${#headers[@]} - 1) ))in" \
+                              "set output \"${LOG_DIR}/${plotname}.pdf\"" \
+                              "set datafile separator \";\"" \    
+                              "set xdata time" \
+                              "set timefmt '%Y-%m-%d %H:%M:%S UTC'" \
+                              "set tmargin 1" \
+                              "set bmargin 1" \
+                              "set lmargin 20" \
+                              "set rmargin 20" \
+                              "set multiplot layout ${#headers[@]},1 title \"\n${plotname}\n\"" \
+                              "unset title" )
+
+    local i
+    for (( i = 1; i < "${#headers[@]}"; i++ )); do
+        local header
+        header="${headers[$i]}"
+
+        if (( $i == ${#headers[@]} - 1 )); then
+            # we need x-tick labels on the bottom plot
+            gnuplot_directive+=( "set xtics format '%H:%M:%S' rotate by -90" )
+        else 
+            gnuplot_directive+=( "set format x ''" )
+        fi 
+
+        gnuplot_directive+=( "plot \"${datafile}\" using 1:$(( $i + 1 )) title \"${header}\" with lines" )
+    done
+
+    # concatenate the array with newlines to get the final directive to send to gnuplot
+    gnuplot_directive="$( IFS=$'\n'; echo "${gnuplot_directive[*]}" )"
+
+    echo -e "\$ gnuplot <<< ${gnuplot_directive}\n" >> "${LOG_DIR}/gnuplot.log"
+    gnuplot <<< "${gnuplot_directive}" >> "${LOG_DIR}/gnuplot.log" 2>&1
+    echo -e "\n\n" >> "${LOG_DIR}/gnuplot.log"
+
+    echo "[INFO] Stacked plot for log subset \"${plotname}\" written to ${LOG_DIR}/${plotname}.pdf"
+}
+
+
+# os::log::install_system_logger installs the system logger and begins logging
+#
+# Globals:
+#  - LOG_DIR
+# Arguments:
+#  None
+# Returns:
+#  - export LOGGER_PID
+#  - export SAR_LOGFILE
+function os::log::install_system_logger() {
+    readonly SAR_LOGFILE="${LOG_DIR}/sar.log"
+    export SAR_LOGFILE
+
+    ( os::log::internal::run_system_logger "${SAR_LOGFILE}" "${LOG_DIR}/sar_stderr.log" ) &
```

Comment:
check for existence of `sar` before trying to use it.  Don't fail on not having it, but echo a message.